ops/pebble.py

```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                    self._headers = None
+                else:
+                    # Data present but boundary not yet found.
+                    # Write data from buffer, but don't accidentally write data which may belong
+                    # to the next boundary line.
+                    safe_bound = max(0, len(self._buffer) - len(self._terminal_boundary))
+                    data = self._buffer[:safe_bound]
+
+                    # Don't open/close file for intermediate writes
+                    outfile = self._get_open_tempfile()
+                    outfile.write(data)
+
+                    # Advance the buffer for bytes consumed
+                    self._buffer = self._buffer[safe_bound:]
+
+                    # Insufficient data to continue
+                    break
+
+    def _parse_headers(self, data):
+        parser = email.parser.BytesFeedParser()
+        parser.feed(data)
+        message = parser.close()
+        return message
+
+    def _detect_invalid_content_disposition(self):
+        content_disposition = self._headers.get_content_disposition()
+        if content_disposition != 'form-data':
+            raise RuntimeError(
+                'unexpected content disposition: {}'.format(content_disposition))
+
+    def _prepare_for_part_body(self):
+        name = self._headers.get_param('name', header='content-disposition')
+        if name == 'response':
+            part_type = 'response'
+        elif name == 'files':
+            filename = self._headers.get_filename()
+            if filename is None:
+                raise NotImplementedError('multipart "files" part missing filename')
+            self._prepare_tempfile(filename)
+            part_type = 'files'
+        else:
+            raise RuntimeError(
+                'unexpected name in content-disposition header: {}'.format(name))
+        return part_type
+
+    def _get_next_boundary_index(self) -> typing.Optional[int]:
+        boundary_index = self._buffer.find(self._boundary_prefix)
+        if self._buffer[boundary_index:].startswith(self._boundary):
+            return boundary_index
+        elif self._buffer[boundary_index:].startswith(self._terminal_boundary):
+            return boundary_index
+        return None
+
+    def _prepare_tempfile(self, filename):
+        tf = tempfile.NamedTemporaryFile(delete=False)
+        self._files[filename] = tf
+        self.current_filename = filename
+
+    def _get_open_tempfile(self):
+        return self._files[self.current_filename]
+
+    def get_response(self):
+        """Return the deserialized JSON object from the multipart "response" field."""
+        return self._response
+
+    def filenames(self):
+        """Return a list of filenames from the "files" parts of the response."""
+        return list(self._files.keys())
+
+    def get_file(self, path, encoding):
+        """Return an open file object containing the data of the specified file."""
```

Comment:
"containing the data"? Probably a leftover.

Reply:
fixed.